[IV-24-25] Objetivo 9

hossam1522 opened the pull request:
[IV-24-25] Objetivo 9

hossam1522 committed:
En relación a #67, me he percatado de que no tengo
forma de acceder al inventario con sentido a la hora de utilizar la API REST. Creo issue

hossam1522 committed:
En relación a #69, se ha pensado que una
opción factible es implementar una estructura que simule una base de datos que incluya
el stock y las ventas

hossam1522 committed:
En relación a #70, se ha creado un test que
comprueba que se pueda insertar ropa correctamente, pero faltan funciones para que
vaya correctamente

hossam1522 committed:
En relación a #70 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/ebbbe3d2428ba2e5dcca5880f7aa2584c24ab37f,
se ha implementado la función para pasar dicho test

hossam1522 committed:
En relación a #70, se ha creado un test que
comprueba que se ha eliminado una prenda, pero falta implementar la función que elimina dicha prenda

hossam1522 committed:
En relación a #69 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/69e31e8ecae88bed2ba34d65b9d531ed7cad3e94,
se ha implementado la función que elimina ropa de la bd (solo elimina una unidad, no tiene sentido, en principio, eliminar más) y se ha modificado una función de
Stock la cual devolvía la ropa aunque la cantidad de esta fuera 0, así que se ha retocado para arreglarlo y pasar los tests correctamente

hossam1522 committed:
En relación a #70, más
tests implementados

hossam1522 committed:
En relación a #70, se empiezan a
implementar los tests relacionados con las ventas. Fallan por falta de funciones

hossam1522 committed:
En relación a #69 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/3c6112f2c3fba959f7057237b79dc1bf65b37ad7,
implementada función necesaria para pasar los tests, la cual incluye la venta en la BD y reduce el stock según sea necesario

hossam1522 committed:
En relación a #70, test implementado
para realizar ventas indicando una fecha. Función no adaptada

hossam1522 committed:
En relación a #69 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/69adfb08fae29f789d9b287bc6c2ce0c65f0748e,
modificada la función para tener en cuenta la fecha

hossam1522 committed:
En relación a #70, se ha implementado
un test para ver si se obtienen las ventas. Falla por falta de funciones

hossam1522 committed:
En relación a #69 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/50fbca4c1fec8be981691225973f76173598f5d3,
se ha modificado el test para que tenga sentido y cuadre con las rutas de la API además de implementar la función que lo soluciona

hossam1522 committed:
En relación a #70, otro
test de obtener ventas de un tipo de prenda pero con fecha

hossam1522 committed:
En relación a #69 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/196604a16a832f0b76e83c9599fc9e5e13f59aae,
modificada función para tener en cuenta también este caso

hossam1522 committed:
En relación a #70, se ha implementado
un test que elimina una venta. Falla por falta de implementacón de la función. closes #70

hossam1522 committed:
En relación a #69 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/740d00a266b0f0b837d256fffb33938dfe279109,
se ha implementado la función que elimina las ventas. closes #69

hossam1522 committed:
En relación a #67, se ha
implementado un test con una BD de prueba para poder implementar la ruta de obtener prendas sin indicar talla

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/91eb369259c80f69ece7db5dbfac26a66c5f4b8f,
se ha tenido que crear un nuevo archivo bd_prueba para poder obtener una base de datos para hacer las pruebas de la API. También me he percatado que para interactuar
con la base de datos van a hacer, mínimo, un par de funciones más

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68/commits/06eb9f8ce34af9d3b39fa44c4ab7dc0d8d271460,
a #69 y a #70, se han implementado dos funciones para recuperar prendas de la bd y se han modificado los tests para que las usen

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/5d9fea6b56307a60fc4cccacad8d82b37f1ceb59,
se ha implementado la función, además de hacer que la resspuesta se devuelva en formato JSON y, además, arreglar la función para que devuelva el router correctamente

hossam1522 committed:
En relación a #67, se han creado
dos tests, uno para comprobar que da error 404 si se solicita una prenda que no existe y otro para solicitar prenda
y talla, pero da el error de que no puede obtener la talla porque no hay método para ello

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68/commits/46887b4c184bc65db6e1a139e84e7b4aa73c5aa7,
se han implementado unos setters en Ropa para poder proceser con los tests de #67, pero por ello me he percatado de que la transformación a JSON no estaba
siendo efectiva

hossam1522 committed:
Primeras pruebas no efectivas para resolver #71

hossam1522 committed:
En relación a #71, se ha conseguido que los
datos se puedan interpretar en formato JSON manteniendo los atributos privados mediante la implementación de
funciones marshal y unmarshal para la estructura de Ropa. closes #71

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/46887b4c184bc65db6e1a139e84e7b4aa73c5aa7,
se ha implementado la ruta para obtener una prenda por nobre y talla

hossam1522 committed:
EN relación a #67, se ha creado un test que
hace un post de un recurso en una determinada cantidad. No funciona debido a la falta de implementación del método

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/ad236a107ef9f070838f935f3fa21e254f20089d,
se ha implementado el post de ropa, además de hacer que la base de datos de prueba sea única para todos los métodos y no se esté generando cada vez que se llama
a un método, para poder hacer las pruebas correctamente

hossam1522 committed:
En relación a #67, se ha implementado un test que
borra una prenda de la base de datos y comprueba si está borrada (como solo hay un ejemplar de la prenda, compruebo que desaparezca a la hora de querer obtenerlo

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/274dc695a02b71e8a8d2f5f8c7f552ab42c460f6,
se ha implementado la rua que hace esto sin problema

hossam1522 committed:
En relación a #67, se ha creado un
test que devuelva las ventas asociada a una prenda y se ha modificado la base de datos para poder realizar correctamente esta prueba

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/68df354e8dc7262efd2fdb0935c596d073e75d67,
se ha implementado la función correspondiente a la ruta y se ha modificado también una función de estructura de BD y uno de los tests para tener en cuenta errores,
pero se produce el mismo error descrito en #71

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68/commits/3c6119eb6c3fb8aae0a713fababd6bdd1da6beb5,
el error no era el mismo que en #71

hossam1522 committed:
En relación a #67 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/68df354e8dc7262efd2fdb0935c596d073e75d67,
se han ampliado las comprobaciones del test y se ve que ocurre el mismo fallo que en #71

hossam1522 committed:
En relación a #71 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/fdb235b1624614216860e8f31d91c7df51124e98,
se ha añadido un marshal y unmarshal a la estructura de venta y ahora se realiza correctamente el test. closes #71

hossam1522 committed:
En relación a #67, se ha creado un
un test que devuelva las ventas asociadas a una prenda en una fecha determinada, añadiendo también ejemplos a la BD de prueba para
poder realizar los tests

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/5586a3c00dfd0953303963094c259a3e0f2bb74d,
implementada correctamente la ruta para realizar el GET, además de cambiado el ejemplo con el que se hace el test para no afectar a otros tests

hossam1522 committed:
En relación a #67, se ha creado un test
que realice una venta y comprueba si se ha realizado y si se ha borrado del inventario la prenda vendida, ya que solo hay una

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/e76bbc8cb4dbadcc4cac8a9e74c1f5c098a61b5a,
se ha implementado correctamente la función de PUT venta

hossam1522 committed:
En relación a #67, se ha creado un test
que elimine una venta. El enfoque seguido es que la venta simplemente se elimina de la base de datos, no se realiza nada más, ya que siguiendo #4
se eliminan ventas pasadas pero ya está, nada más. closes #67

hossam1522 committed:
En relación a #66 y a https://github.com/hossam1522/ModaTrack/pull/68/commits/622a6affd46404c14b529fd81321871cacb85312,
se ha implementado el DELETE de la API y funciona correctamente. closes #66

hossam1522 committed:
En relación a #72, añadida
una task al taskfile para compilar el proyecto usando build. closes #72

JJ reviewed (changes requested):
El principal problema es que tienes demasiada repetición, posiblemente resultado de usar IAs generativas. A estas alturas no vamos a decir aquello de "Diantre, en ese local se juega", pero sí sería conveniente que examinaras de forma crítica el código generado y usaras tu inteligencia natural general para refactorizar para obtener código que siga con más precisión los dictámenes del código limpio, así como aplicar ciertos patrones un poco más avanzados.

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903092986,
se ha quitado la función y ruta para eliminar una prenda por no venir incluida en las historias de usuario

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903092137,
se han eliminado los números mágicos haciendo los test menos frágiles

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903092488 y a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903092865, se han establecido mensajes de error más informativos

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903093178 y a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903093611, modificada la estructura y todas las llamadas a la base de datos para usar inyección de dependencias y dejar el código claro y consistente

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903092968,
eliminado el código repetido a la hora de pillar los parámetros de la ruta añadiendo un middleware que coge del contexto los parámetros,
los asigna a un map, este map se añade al contexto y luego se pueden coger los parámetros directamente de este contexto sin necesidad
de estar declarando los parámetros explícitamente para trabajar con ellos haciendo uso de chi.URLParam

JJ reviewed (changes requested):
Es posible que tengas que inyectar tanto el enrutador como el almacén de datos. En todo caso, tener algo llamado `getBDPrueba` en código de producción no parece una buena práctica, ¿no te parece?

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903770622, https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903771245 y a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903771644, se han añadido más detalles a los mensajes de error además de hacer que estos detalles se puedan leer sin problema si ocurre algún fallo en las pruebas de api_test

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1903773910,
usada la inyección de dependencias para evitar tener código de prueba en producción

JJ reviewed (changes requested):
El principal problema es que no acabas de solucionar bien lo de la BD de prueba, ni desde el punto de vista de separar código de producción del de prueba ni desde el punto de vista de inyectar dependencias. Mira a ver si [esto te ayuda](https://stackoverflow.com/a/56179521/891440) al menos para crear diferentes constructores o un constructor con diferentes opciones para la BD. Mira también [esta solución](https://www.reddit.com/r/golang/comments/wbawx5/comment/ii5m2ox/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) para inyección de dependencias en Go, pero hay más en ese mismo hilo.

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1904494758,
eliminados numeros mágicos relacionados con códigos de error

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1904495426,
eliminada la ruta para eliminar una venta por no venir del todo bien contemplada en las HUs

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1904495767,
eliminado código repetido para evitar utilizar mucho el mismo patrón

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1904498161,
se ha eliminado el código de prueba de producción y se ha creado un JSON para tests

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1904500716,
implementado un constructor para la BD que inicialice desde un fichero, usando como guía lo especificado en https://stackoverflow.com/a/56179521/891440,
eliminando definitivamente una BD Prueba del código de producción

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#pullrequestreview-2532728044,
arreglado, en principio, el problema de no estar usando correctamente la inyección de dependencias

JJ reviewed (approved):
La refactorización no acaba de seguir las buenas prácticas, sigue habiendo varios getters que probablemente no necesites, pero el objetivo está alcanzado.

hossam1522 committed:
En relación a https://github.com/hossam1522/ModaTrack/pull/68\#discussion_r1904983695,
cambiado el argumento de las rutas de  a  al solo estarse usando la bd
